flocker/testtools/__init__.py

```diff
@@ -609,7 +613,11 @@ def build(self, dockerfile_variables=None):
             docker_dir.path
         ]
         run_process(command)
-        self.test.addCleanup(run_process, [b"docker", b"rmi", tag])
+        if self.cleanup:
+            def remove_image():
+                client = DockerClient()
+                client.remove_image(tag, force=True)
```

Comment:
If this removed all containers using this image and then removed the image, the result would probably be more useful.  This would let all users of `DockerImageBuilder` (with `cleanup=True`) ignore container cleanup and have the right thing happen.  It would also avoid leaving dangling broken containers on the system.

File a follow-up issue and refer to it from here if you don't want to implement this behavior now.